flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java

```diff
@@ -37,9 +39,12 @@
 
 	private final int priority;
 
-	public MailboxExecutorImpl(@Nonnull TaskMailbox mailbox, int priority) {
+	public final StreamTaskActionExecutor actionExecutor;
```

Comment:
I might have misunderstood the full intent of the ticket, but isn't it the main idea to simplify the code path of all tasks, which are not legacy sources?
This line introduces a new indirection for executing a mail for _all_ tasks. 
I'd rather see two different MailboxExecutor implementations, one for legacy and one for everything else (=as is). Alternatively, we could also have decorator around the normal MailboxExecutor that synchronizes.

Reply:
> Why is this public?

Fixed, thanks.


The intent of this PR is to:
1. make use of mailbox executor safe without the `checkpoint lock`
2. prepare for push `checkpoint lock` down to `SourceStreamTask`

Having a custom mailbox processor in `SourceStreamTask` would still require to synchronize some actions in `StreamTask`.
As for performance, `actionExecutor` is not used for the hot path, i.e. `default action`.
(and I'd love to see any data on how such a design (method call stack of depth 2 instead of 1) affects Flink performance)